### PR TITLE
save and restore forms by js

### PR DESCRIPTION
**Pull request by orthagh**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

I was frustrated yesterday by losing a followup on support.
The tab was open for a long time and so i hit expiracy of the current csrf token.

So this feature/fix try to save live all actions done on inputs/tinyme (no others fields) of a ticket to retore them on a page reload (after a crash, etc) like done here in github comment.

PR is not finished, some points to adress:

- [ ] expand to change/problem (i didn't find a selector to diferenciate from a ticket)
- [ ] remove storage after a successful post (i didn't though intensively about that, maybe using localStorage is not a good idea)
- [ ] future maintenance of this script. Ex, i'll check later, but i think the restore function of a followup template is currently broken

But i wanted to get a first round of opinions before going further

### CODE REVIEW

**Comment by cconard96** [comment]
Is there any concern that local storage is readable by any script on any page in the same domain? Like if one GLPI page has a XSS flaw, a malicious script could read what was entered in ITIL forms last (especially if it isn't cleared).

When people switch users, there may also be a risk of restoring content from another user session. Worse case I can think of is a shared computer where a technician may log in to GLPI, update someone's ticket, and then log out of GLPI. The next user could manually view all the content in the local storage, or have it accidentally restored when they try filling out a form.

**Comment by orthagh** [comment]
> Is there any concern that local storage is readable by any script on any page in the same domain? Like if one GLPI page has a XSS flaw, a malicious script could read what was entered in ITIL forms last (especially if it isn't cleared).

Good point, i may switch to IndexedDB

> When people switch users, there may also be a risk of restoring content from another user session. Worse case I can think of is a shared computer where a technician may log in to GLPI, update someone's ticket, and then log out of GLPI. The next user could manually view all the content in the local storage, or have it accidentally restored when they try filling out a form.

I'll think about that

**Comment by orthagh** [comment]
I can't find where github store their data. It seems to be client side as i don't see any ajax request, but all storage doesn't seems to be updated

**Comment by orthagh** [comment]
For github: https://github.com/github/session-resume

They use SessionStorage, and combine it with pageshow and pagehide event to move from the storage to a somehow global

**Comment by cconard96** [comment]
So, do nothing unless the page is hidden. When that page is hidden, the fields are stored in the session storage client-side. When the page is shown again, the fields are restored and removed from session storage.

I guess that makes it a bit more secure.

If it only saves and restores using the page events though, I wonder if the data be saved on the server instead like it currently is by using an asynchronous request during the pagehide event and then either letting the server populate the fields like it already does, or ask the server for them later with another AJAX request.

File: src/Html.php [hunk 1]
```diff
@@ -6451,9 +6458,14 @@ public static function getCoreVariablesForJavascript(bool $full = false)
         }
         $plugins_path = 'var GLPI_PLUGINS_PATH = ' . json_encode($plugins_path) . ';';
 
+        $session_glpi = "var SESSION_GLPI = {
```
Comment (orthagh) [inline src/Html.php:6461]: I gladly take opinions on this particular exposition

Comment (cedric-anne) [reply src/Html.php:6461]: I am not sure it is a good idea to expose user ID, even if it is probably already exposed by debug bar information.
Maybe a name-based UID would permit to have a stable unique identifier for a given user, without exposing sensitive information.
```
\Ramsey\Uuid\Uuid::uuid5(\Ramsey\Uuid\Uuid::NIL, sprintf('%d-%s', $_SESSION["glpiID"], $_SESSION["glpiname"]));
```

**Review by cedric-anne — COMMENTED** [review]
See comment.

**Pull request closed without merging**